File: CivitasJava/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author luis
 */
public class JugadorEspeculador extends Jugador {
    private static int FactorEspeculador = 2;
    private float fianza;
    
    JugadorEspeculador(Jugador jugador, float fianza){
        super(jugador);
        this.fianza = fianza;
    }
    
    private void actualizaPropietarioPorConversion(){
        for(int i = 0; i < this.propiedades.size(); i++){
            propiedades.get(i).actualizaPropietarioPorConversion(this);
        }                  
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado(numCasillaCarcel)){
            if(this.saldo > fianza){
                this.paga(fianza);
                Diario diary = Diario.getInstance();
                diary.ocurreEvento("Jugador: " + this.getNombre() + " no es encarcelado");
                encarcelado = false;
            }else{
                moverACasilla(numCasillaCarcel);
                encarcelado = true;
                Diario diary = Diario.getInstance();
                diary.ocurreEvento("Jugador: " + this.getNombre() + " encarcelado");
            }
        }
        
        return encarcelado;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        if(isEncarcelado()){
            return false;
        }else{
            float cantidadEspeculador = cantidad/this.FactorEspeculador; 
            return paga(cantidadEspeculador);
        }
    }
    
    @Override
    int getCasasMax(){
        return this.casaMax * FactorEspeculador;
    }
    
    @Override
    int getHotelesMax(){
        return hotelesMax * FactorEspeculador;
    }
    
    @Override
    public String toString(){
        String e = "Jugador Especulador: " + this.getNombre() + "\nSaldo: " + this.getSaldo() + "\nCasilla: " + this.getNumCasillaActual() + "\nEncarcelado: " + this.isEncarcelado() + "\nFianza: " +  this.fianza;
        return e;
    }
    
    
}

File: CivitasRuby/lib/jugador_especulador.rb
# encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class JugadorEspeculador < Jugador
    
    @@FactorEspeculador = 2
    
    def initialize(jugador,fianza)
      @fianza = fianza
      super(jugador)
    end
    
    
    def actualiza_propietario_por_conversion()
      for i in self.propiedades
        i.actualiza_propietario_por_conversion(self)
      end
    end
    
    def self.new_Especulador(jugador, fianza)
      jugador_especulador = new
      jugador_especulador.actualiza_propietario_por_conversion()
      return jugador_especulador
    end
    
    def encarcelar(numCasillaCarcel)
      if debe_ser_encarcelado(numCasillaCarcel)
        if self.saldo > @fianza
          self.paga(@fianza)
          @encarcelado = false
          diary = Diario.instance
          diary.ocurre_evento("Jugador: #{@nombre} no es encarcelado")
        else
          mover_a_casilla(numCasillaCarcel)
          @encarcelado = true
          diary = Diario.instance
          diary.ocurre_evento("Jugador: #{@nombre} encarcelado")
        end
        
      end
      
      return @encarcelado
    end
    
    def paga_impuesto(cantidad)
      if is_encarcelado
        return false
      else
        cantidad_especulador = cantidad/@@FactorEspeculador
        return paga(cantidad_especulador)
      end
    end
    
    def casaMax
      return self.casaMax * @@FactorEspeculador
    end
    
    def hotelesMax
      return self.hotelesMax * @@FactorEspeculador
    end
    
    def to_s
      "Jugador Especulador:  #{@nombre.to_s}  \n Saldo:  #{@saldo.to_s}  \nCasilla:   #{@numCasillaActual.to_s}  \nEncarcelado:  #{@encarcelado.to_s}"
    end
    
    
    #private_class_method :new
  end
end
